architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,442 @@
+| OpenAPI/Schema                | A format for documenting HTTP-based APIs. We use OpenAPI v2 currently in Radius for client-sdk generation and validation of inbound requests. OpenAPI v3 is more widely adopted. In this document we're primarily concerned with the parts of OpenAPI pertaining to request/response bodies.                                                                                                                                                                                                                          |
+| Declarative User-Defined Type | An implementation of User-Defined Types that relies on delarative functionality like schemas for validation, and recipes for provisioning. This is a new model for defining types in Radius and is the focus of this document. All existing Radius functionality implements resource provider microservices (not declarative).                                                                                                                                                                                        |
+| Tracked resource              | A resource type that is *tracked* by UCP. Tracked resources are cached by UCP and their existence can easily be queried without sending a request to the resource provider. All existing resource types in Radius are tracked resources.                                                                                                                                                                                                                                                                              |
+| Normal resource type          | A resource type that is part of the `radius` plane and exists inside a resource group. Also has the `TrackedResource` set of fields and behaviors. This is the case for most resource types in Radius today. Normal resource types are our scenario priority because they are optimized for being part of an application. The resource types that are not *normal* are not usually part of an application, for example `System.AWS/credentials` is not part of an application and does not exist in a resource group. |
+
+## Objectives
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6688
+
+### Guiding principles
+
+We have two guiding principles for how we make decisions and prioritize investments. 
+- User-defined types must be simple and declarative for application and recipe scenarios.  The highest value scenario is building custom application platforms. We want to build a big ecosystem and so user-defined types should be easy to adopt and easy to develop.
+- User-defined types should be powerful enough for the Radius maintainers to use. We want to leverage user-defined types and declarative functionality as much as possible.
+
+
+
+### Goals
```

Comment:
Nicely written goals!

Reply:
Thanks! I'm going for a world record with how many there are 🥇